rangeproof/splitutils.go

```diff
@@ -0,0 +1,84 @@
+package rangeproof
+
+import (
+	"errors"
+
+	"github.com/privacybydesign/gabi/big"
+	"github.com/privacybydesign/gabi/internal/common"
+)
+
+// SquareSplitter provides a combined interface for all facets describing a method for spliting positive numbers into a sum of squares.
+type SquareSplitter interface {
+	// Number of bits per square
+	Ld() uint
+	// Number of squares in result
+	SquareCount() int
+	// Actual splitting function, on input delta, should return array x such that sum_i x_i^2 = delta and len(x) = SquareCount()
+	Split(*big.Int) ([]*big.Int, error)
+}
+
+type SquaresTable [][]int64
+
+// Generate lookup table for splitting numbers into 3 squares containing entries up-to and including limit
+// takes O(n^3/2)
+func GenerateSquaresTable(limit int64) SquaresTable {
```

Comment:
This returns a `SquaresTable`, but `*SquaresTable` is the type that implements the `SquareSplitter` interface, forcing callers to sprinke `&` through their code. Can we change the receiver of the methods below to just `SquaresTable` (if so, then also for `FourSquaresSplitter` for consistency)? If not, perhaps this function can return a `*SquaresTable`.

Reply:
fixed, chosen for pointer type to avoid too much copying